consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/IoTDBMetricRegistry.java

```diff
@@ -0,0 +1,200 @@
+import org.apache.ratis.metrics.RatisMetricRegistry;
+
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class IoTDBMetricRegistry implements RatisMetricRegistry {
+
+  private final AbstractMetricService metricService;
+  private final MetricRegistryInfo info;
+  private final String prefix;
+  private final Map<String, String> metricNameCache = new ConcurrentHashMap<>();
+  private final Map<String, CounterProxy> counterCache = new ConcurrentHashMap<>();
+  private final Map<String, TimerProxy> timerCache = new ConcurrentHashMap<>();
+  private final Map<String, GaugeProxy> gaugeCache = new ConcurrentHashMap<>();
+
+  IoTDBMetricRegistry(MetricRegistryInfo info, AbstractMetricService service) {
+    this.info = info;
+    this.metricService = service;
+    prefix =
+        MetricRegistry.name(
+            info.getApplicationName(), info.getMetricsComponentName(), info.getPrefix());
+  }
+
+  private String getMetricName(String name) {
+    return metricNameCache.computeIfAbsent(name, n -> MetricRegistry.name(prefix, n));
+  }
+
+  @Override
+  public Timer timer(String name) {
+    final String fullName = getMetricName(name);
+    return timerCache.computeIfAbsent(
+        fullName, fn -> new TimerProxy(metricService.getOrCreateTimer(fn, MetricLevel.IMPORTANT)));
+  }
+
+  @Override
+  public Counter counter(String name) {
+    final String fullName = getMetricName(name);
+    return counterCache.computeIfAbsent(
+        fullName,
+        fn ->
+            new CounterProxy(
+                metricService.getOrCreateCounter(getMetricName(name), MetricLevel.IMPORTANT)));
+  }
+
+  @Override
+  public boolean remove(String name) {
+    // TODO (szywilliam) we can add an interface like removeTypeless(name)
+
+    try {
+      metricService.remove(MetricType.COUNTER, getMetricName(name));
+    } catch (IllegalArgumentException ignored) {
+    }
+    try {
+      metricService.remove(MetricType.TIMER, getMetricName(name));
+    } catch (IllegalArgumentException ignored) {
+    }
+    try {
+      metricService.remove(MetricType.AUTO_GAUGE, getMetricName(name));
+    } catch (IllegalArgumentException ignored) {
+    }
+
+    return true;
+  }
+
+  void removeAll() {
+    counterCache.forEach((name, counter) -> metricService.remove(MetricType.COUNTER, name));
+    gaugeCache.forEach((name, gauge) -> metricService.remove(MetricType.AUTO_GAUGE, name));
+    timerCache.forEach((name, timer) -> metricService.remove(MetricType.TIMER, name));
+    metricNameCache.clear();
+    counterCache.clear();
+    gaugeCache.clear();
+    timerCache.clear();
+  }
+
+  @Override
+  public Gauge gauge(String name, MetricRegistry.MetricSupplier<Gauge> metricSupplier) {
+    final String fullName = getMetricName(name);
+    final GaugeProxy gauge = new GaugeProxy(metricSupplier);
+    final GaugeProxy previous = gaugeCache.putIfAbsent(fullName, gauge);
+    if (previous == null) {
```

Comment:
Please ensure atomicity.

Reply:
Thanks for pointing out this! Will switch to computeIfAbsent